csp/handlers/handlers.go

```diff
@@ -0,0 +1,251 @@
+	} else if len(req.Email) > 0 && !internal.ValidEmail(req.Email) {
+		return nil, errors.ErrInvalidUserData.Withf("invalid email format")
+	} else if len(req.Phone) > 0 {
+		var err error
+		if req.Phone, err = internal.SanitizeAndVerifyPhoneNumber(req.Phone); err != nil {
+			log.Warnw("invalid phone number format", "phone", req.Phone, "error", err)
+			return nil, errors.ErrInvalidUserData.Withf("invalid phone number format")
+		}
+	}
+	// get census information
+	census, err := c.mainDB.Census(censusID)
+	if err != nil {
+		if err == db.ErrNotFound {
+			return nil, errors.ErrCensusNotFound
+		}
+		return nil, errors.ErrGenericInternalServerError.WithErr(err)
+	}
+	// check the census membership of the participant
+	if _, err := c.mainDB.CensusMembership(censusID, req.ParticipantNo); err != nil {
+		if err == db.ErrNotFound {
+			return nil, errors.ErrUnauthorized.Withf("participant not found in the census")
+		}
+		return nil, errors.ErrGenericInternalServerError.WithErr(err)
+	}
+	// get participant information
+	participant, err := c.mainDB.OrgParticipantByNo(census.OrgAddress, req.ParticipantNo)
+	if err != nil {
+		return nil, errors.ErrCensusParticipantNotFound
+	}
+	// check the password
+	if req.Password != "" && !bytes.Equal(internal.HashPassword(c.csp.PasswordSalt, req.Password), participant.HashedPass) {
+		return nil, fmt.Errorf("invalid user data")
+	}
+	// check the census type and parse the destination and notification type
+	// accordingly
+	var toDestinations string
+	var challengeType notifications.ChallengeType
+	switch census.Type {
+	case db.CensusTypeMail:
+		if !bytes.Equal(internal.HashOrgData(census.OrgAddress, req.Email), participant.HashedEmail) {
+			return nil, errors.ErrUnauthorized.Withf("invalid user email")
+		}
+		toDestinations = req.Email
+		challengeType = notifications.EmailChallenge
+	case db.CensusTypeSMS:
+		if !bytes.Equal(internal.HashOrgData(census.OrgAddress, req.Phone), participant.HashedPhone) {
+			return nil, errors.ErrUnauthorized.Withf("invalid user phone")
+		}
+		toDestinations = req.Phone
+		challengeType = notifications.SMSChallenge
+	case db.CensusTypeSMSorMail:
+		if req.Email != "" {
+			if !bytes.Equal(internal.HashOrgData(census.OrgAddress, req.Email), participant.HashedEmail) {
+				return nil, errors.ErrUnauthorized.Withf("invalid user email")
+			}
+			toDestinations = req.Email
+			challengeType = notifications.EmailChallenge
+		} else if req.Phone != "" {
+			if !bytes.Equal(internal.HashOrgData(census.OrgAddress, req.Phone), participant.HashedPhone) {
+				return nil, errors.ErrUnauthorized.Withf("invalid user phone")
+			}
+			toDestinations = req.Phone
+			challengeType = notifications.SMSChallenge
+		}
+	default:
+		return nil, errors.ErrNotSupported.Withf("invalid census type")
+	}
+	// parse the user ID and generate the token
+	userID := new(internal.HexBytes).SetString(participant.ID.Hex())
```

Comment:
@lucasmenendez this is incosistent with the way the userID is generated when creating the process:
https://github.com/vocdoni/saas-backend/blob/8c887c0f6a82971cd7d4eed5a184034e24732790/api/process_bundles.go#L132